src/daft-csv/src/read.rs

```diff
@@ -190,119 +289,1081 @@ where
             Series::try_from((field.name.as_ref(), cast_array_for_daft_if_needed(array)))
         })
         .collect::<DaftResult<Vec<Series>>>()?;
+    // Build Daft Table.
     let schema: arrow2::datatypes::Schema = fields.into();
     let daft_schema = Schema::try_from(&schema)?;
     Table::new(daft_schema, columns_series)
 }
 
+async fn read_into_column_chunks<R>(
+    mut reader: AsyncReader<Compat<R>>,
+    fields: Arc<Vec<arrow2::datatypes::Field>>,
+    projection_indices: Arc<Vec<usize>>,
+    num_rows: Option<usize>,
+    chunk_size: usize,
+    estimated_mean_row_size: usize,
+    max_chunks_in_flight: usize,
+) -> DaftResult<Vec<Vec<Box<dyn arrow2::array::Array>>>>
+where
+    R: AsyncRead + Unpin + Send,
+{
+    let num_fields = fields.len();
+    let num_rows = num_rows.unwrap_or(usize::MAX);
+    let chunk_size_rows = {
+        // With a default chunk size of 64 KiB and estimated bytes per row of 200 bytes,
+        // this would result in a chunk size of ~328 rows.
+        let estimated_rows_per_desired_chunk = chunk_size / estimated_mean_row_size;
+        // Process at least 8 rows in a chunk, even if the rows are pretty large.
+        estimated_rows_per_desired_chunk.max(8).min(num_rows)
+    };
+    // Stream of unparsed CSV byte record chunks.
+    let read_stream = async_stream::try_stream! {
+        // Number of rows read in last read.
+        let mut rows_read = 1;
+        // Total number of rows read across all reads.
+        let mut total_rows_read = 0;
+        while rows_read > 0 && total_rows_read < num_rows {
```

Comment:
What about computing `estimated_mean_row_size` as a running average of `bytes_read / rows_read `

Reply:
Working on that now!